src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+                StartResult::FailedStart {
+                    failed_worker_spec,
+                    start_error,
+                    prev_started_workers,
+                    pending_workers,
+                }
+            }
+        }
+    }
+}
+
+async fn start_children<'a, 'b>(
+    ctx: &'a Context,
+    mut ev_notifier: EventNotifier,
+    sup_runtime_name: &str,
+    sup_order: &'b Order,
+    // TODO: receive error notifier parameter
+    children_specs0: Vec<worker::Spec>,
+) -> Result<Vec<Worker>, (Vec<worker::Spec>, StartError)> {
+    let mut start_result = StartResult::new(children_specs0.len());
+    let pending_children: Box<dyn Iterator<Item = worker::Spec>> =
+        ordered_children_on_start(sup_order, children_specs0);
+
+    for failed_child_spec in pending_children {
+        start_result = start_result
+            .start_child(
+                ctx,
+                ev_notifier.clone(),
+                sup_runtime_name,
+                failed_child_spec,
+            )
+            .await;
+    }
+
+    match start_result {
+        StartResult::SuccessfulStart {
+            started_workers, ..
+        } => Ok(started_workers),
+        StartResult::FailedStart {
+            failed_worker_spec,
+            start_error,
+            prev_started_workers,
+            pending_workers,
+        } => {
+            let termination_result = terminate_prev_siblings(
+                &mut ev_notifier,
+                failed_worker_spec,
+                start_error,
+                pending_workers,
+                prev_started_workers,
+            )
+            .await;
+            Err(termination_result)
+        }
+    }
+}
+
+// terminate_children terminates each worker::Worker record. In case any of the
+// children fail to terminate a supervisor::TerminationError is returned.
+//
+async fn terminate_children<'a>(
+    ev_notifier: &mut EventNotifier,
+    runtime_children: impl Iterator<Item = Worker>,
+) -> Result<Vec<worker::Spec>, (Vec<worker::Spec>, TerminationError)> {
+    let mut children_specs = Vec::new();
+    let mut termination_errors: HashMap<String, Arc<worker::TerminationError>> = HashMap::new();
+    for runtime_child in runtime_children {
+        let child_runtime_name = runtime_child.runtime_name.to_owned();
+        let result = runtime_child.terminate().await;
+        match result {
+            Err((failed_child_spec, termination_err0)) => {
+                let termination_err = Arc::new(termination_err0);
+                ev_notifier
+                    .worker_termination_failed(child_runtime_name.clone(), termination_err.clone())
+                    .await;
+                termination_errors.insert(child_runtime_name, termination_err);
+                children_specs.push(failed_child_spec);
+            }
+            Ok(failed_child_spec) => {
+                ev_notifier
+                    .worker_terminated(child_runtime_name.clone())
```

Comment:
You don't need to clone the `child_runtime_name` on this branch since it isn't used twice.